test/src/loaders/workspace_uploader/workspace_uploader_test.py

```diff
@@ -223,78 +237,220 @@ def test_post_process(setup_and_teardown):
-    assembly_dirs = params.assembly_dirs
+    output_dir = Path(params.tmp_dir) / "output_dir"
+    output_dir.mkdir()
 
     wait_to_upload_assemblies = {
         assembly_name: assembly_dir
         for assembly_name, assembly_dir in zip(ASSEMBLY_NAMES, assembly_dirs)
     }
 
-    conf = Mock()
-    conf.workers = 5
-    conf.input_queue = Queue()
-    conf.output_queue = Queue()
-
-    # an uploaded successful
-    conf.output_queue.put((ASSEMBLY_NAMES[0], "12345_58_1"))
-    # an upload failed
-    conf.output_queue.put((ASSEMBLY_NAMES[1], None))
-
-    failed_names = workspace_uploader._upload_assembly_files_in_parallel(
-        conf, "CI", 12345, upload_dir, wait_to_upload_assemblies
-    )
+    ws = create_autospec(Workspace, spec_set=True, instance=True)
+    asu = create_autospec(AssemblyUtil, spec_set=True, instance=True)
+    asu.save_assemblies_from_fastas.return_value = {
+        "results":[
+            {"upa": "12345/58/1"},
+            {"upa": "12345/60/1"}
+        ]
+    }
 
-    expected_tuple1 = (
+    uploaded_count = workspace_uploader._upload_assembly_files_in_parallel(
+        asu,
+        ws,
         "CI",
         12345,
-        os.path.join(
-            workspace_uploader.JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, ASSEMBLY_NAMES[0]
-        ),
-        assembly_dirs[0],
-        ASSEMBLY_NAMES[0],
+        "214",
         upload_dir,
-        1,
-        len(ASSEMBLY_NAMES),
+        wait_to_upload_assemblies,
+        2,
+        output_dir,
     )
 
-    expected_tuple2 = (
+    assert uploaded_count == 2
+
+    # check softlink for post_process
+    assert os.readlink(os.path.join(upload_dir, "12345_58_1")) == os.path.join(
+        output_dir, "12345_58_1"
+    )
+    assert os.readlink(os.path.join(upload_dir, "12345_60_1")) == os.path.join(
+        output_dir, "12345_60_1"
+    )
+
+    # check hardlink for post_process
+    assert os.path.samefile(
+        src_files[0], os.path.join(os.path.join(output_dir, "12345_58_1"), f"12345_58_1.fna.gz")
+    )
+
+    assert os.path.samefile(
+        src_files[1], os.path.join(os.path.join(output_dir, "12345_60_1"), f"12345_60_1.fna.gz")
+    )
+
+
+def test_fail_upload_assembly_files_in_parallel(setup_and_teardown):
+    params = setup_and_teardown
+    assembly_dirs = params.assembly_dirs
+    upload_dir = Path(params.tmp_dir) / "upload_dir"
+    upload_dir.mkdir()
+    output_dir = Path(params.tmp_dir) / "output_dir"
+    output_dir.mkdir()
+
+    wait_to_upload_assemblies = {
+        assembly_name: assembly_dir
+        for assembly_name, assembly_dir in zip(ASSEMBLY_NAMES, assembly_dirs)
+    }
+
+    ws = create_autospec(Workspace, spec_set=True, instance=True)
+    asu = create_autospec(AssemblyUtil, spec_set=True, instance=True)
+    asu.save_assemblies_from_fastas.side_effect = Exception("Illegal character in object name")
+
+    loader_helper.list_objects = create_autospec(loader_helper.list_objects)
+    loader_helper.list_objects.return_value = []
```

Comment:
Is there any reason to do it this way rather than just setting the return value on ws? Monkey patching always makes me nervous - in this case the patch is never reset and could therefore mess up future tests

Reply:
👍

Reply:
:+1: